lib/assert.js

```diff
@@ -931,6 +932,28 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+// Must be called before assert.mustSettle can be used
+assert.initMustSettle = function initMustSettle() {
+  createHook({
+    init(id, type) {}
+  }).enable();
+};
+
+// If promise has not settled before process exit
+// Process exits with non zero exit code
+assert.mustSettle = function mustSettle(promise) {
+  if (checkIsPromise(promise)) {
+    process.on('exit', () => {
```

Comment:
Registering a separate exit handler for every Promise is going to be problematic. I'd recommend taking a look at how things are done for `common.mustCall()`. There is a single process.exit handler that works through the full bucket of functions that have been passed to `common.mustCall()`. Something like that should be done here.

Also, once a Promise does settle, there's no reason it should be kept around, but registering the on('exit') handler like this will prevent it from being garbage collected and will result in a massive memory leak. So, unfortunately, this approach is not going to work.

Reply:
Ah yes, I didn't considered the memory leak this would cause. I'll have a look at `common.mustCall()`